src/Locator/RootsFileOrDirectoryLocator.php

```diff
@@ -33,66 +33,89 @@
 
 declare(strict_types=1);
 
-namespace Infection\Finder;
+namespace Infection\Locator;
 
-use Infection\Finder\Exception\LocatorException;
+use function current;
+use const DIRECTORY_SEPARATOR;
 use function Safe\realpath;
 use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\Assert\Assert;
+use Webmozart\PathUtil\Path;
 
 /**
  * @internal
  */
-final class Locator implements LocatorInterface
+final class RootsFileOrDirectoryLocator implements Locator
 {
-    /**
-     * @var string[]
-     */
-    private $paths;
+    private $roots;
+    private $filesystem;
 
     /**
-     * @var Filesystem
+     * @param  string[] $roots
      */
-    private $filesystem;
-
-    public function __construct(array $paths, Filesystem $filesystem)
+    public function __construct(array $roots, Filesystem $filesystem)
     {
-        $this->paths = $paths;
+        Assert::allString($roots);
+
+        $this->roots = $roots;
         $this->filesystem = $filesystem;
     }
 
-    public function locate(string $name): string
+    /**
+     * {@inheritdoc}
+     */
+    public function locate(string $fileName): string
     {
-        if ($this->filesystem->isAbsolutePath($name)) {
-            if ($this->filesystem->exists($name)) {
-                return realpath($name);
+        $canonicalFileName = Path::canonicalize($fileName);
+
+        if ($this->filesystem->isAbsolutePath($canonicalFileName)) {
+            if ($this->filesystem->exists($canonicalFileName)) {
+                return realpath($canonicalFileName);
             }
 
-            throw LocatorException::fileOrDirectoryDoesNotExist($name);
+            throw FileNotFound::createForFile($canonicalFileName, $this->roots);
         }
 
-        foreach ($this->paths as $path) {
-            $file = $path . \DIRECTORY_SEPARATOR . $name;
+        foreach ($this->roots as $path) {
+            $file = $path . DIRECTORY_SEPARATOR . $canonicalFileName;
 
             if ($this->filesystem->exists($file)) {
                 return realpath($file);
             }
         }
 
-        throw LocatorException::filesOrDirectoriesDoNotExist($name, $this->paths);
+        throw FileNotFound::createForFile($canonicalFileName, $this->roots);
     }
 
+    /**
+     * {@inheritdoc}
+     */
     public function locateOneOf(array $fileNames): string
     {
-        if (!$fileNames) {
-            throw LocatorException::filesNotFound();
+        $file = $this->innerLocateOneOf($fileNames);
+
+        if (null === $file) {
+            throw FileNotFound::createForFiles($fileNames, $this->roots);
+        }
+
+        return $file;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    private function innerLocateOneOf(array $fileNames): ?string
+    {
+        if ([] === $fileNames) {
+            return null;
         }
 
         try {
-            return $this->locate($fileNames[0]);
-        } catch (\Exception $e) {
+            return $this->locate(current($fileNames));
```

Comment:
Personally, I’m not a friend of using a `return` statement in a `try`-block (as I prefer to reduce the scope of it to statements that can actually throw an exception, and `return` isn’t one of it) what do you think about introducing a temporary variable and returning it after the `try`-and-`catch` block?

Reply:
One more variable, one more thing keep track of. That's really taxing. Better without.

Reply:
Like so?

```php
try {
    $path = $this->locate(current($fileNames));
} catch (FileNotFound $exception) {
    // Continue
}

if (isset($path)) {
    return $path;
}

array_shift($fileNames);

return $this->innerLocateOneOf($fileNames);
```

Not any clearer/readable IMO

Reply:
More like

```php
try {
    $path = $this->locate(current($fileNames));
} catch (FileNotFound $exception) {
    // ...
}

return $path;
```

Reply:
But then you are returning in the `catch` block... why is returning in the `try` problematic but not in the `catch`?

Reply:
It’s not problematic to use a `return` statement in a `try`-block, it’s only a personal convention to reduce the scope of a `try`-block to statements that can actually throw an exception, that’s all.

Reply:
Ah I see. It does make sense, but I don't think it really helps out here :)